From e55f5861ac1f9458784fe0f4ddfdd9b62760f0fb Mon Sep 17 00:00:00 2001
From: Frank Lichtenheld <frank@lichtenheld.com>
Date: Mon, 20 Dec 2021 13:36:01 +0100
Subject: [PATCH 6/6] reactive_socket_service_base: add constructor for
 basic_implementation_type
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Otherwise we see compiler warnings with gcc 10 or newer:
asio/detail/impl/epoll_reactor.ipp:228:28: error:
‘<unnamed>.asio::basic_stream_socket<asio::ip::tcp, asio::any_io_executor>::<unnamed>.asio::basic_socket<asio::ip::tcp, asio::any_io_executor>::impl_.asio::detail::io_object_impl<asio::detail::reactive_socket_service<asio::ip::tcp>, asio::any_io_executor>::implementation_.asio::detail::reactive_socket_service<asio::ip::tcp>::implementation_type::<unnamed>.asio::detail::reactive_socket_service_base::base_implementation_type::reactor_data_’
may be used uninitialized [-Werror=maybe-uninitialized]
  228 |   target_descriptor_data = source_descriptor_data;
      |                            ^~~~~~~~~~~~~~~~~~~~~~

Since reactor_data_ can be either a pointer or a empty struct
use the default initializer which does the right thing either
way.

Signed-off-by: Frank Lichtenheld <frank@lichtenheld.com>
---
 include/asio/detail/reactive_socket_service.hpp      | 3 ++-
 include/asio/detail/reactive_socket_service_base.hpp | 8 ++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/include/asio/detail/reactive_socket_service.hpp b/include/asio/detail/reactive_socket_service.hpp
index be84b522e..f21a0de69 100644
--- a/include/asio/detail/reactive_socket_service.hpp
+++ b/include/asio/detail/reactive_socket_service.hpp
@@ -66,7 +66,8 @@ public:
   {
     // Default constructor.
     implementation_type()
-      : protocol_(endpoint_type().protocol())
+      : base_implementation_type{},
+        protocol_(endpoint_type().protocol())
     {
     }
 
diff --git a/include/asio/detail/reactive_socket_service_base.hpp b/include/asio/detail/reactive_socket_service_base.hpp
index 74eed1fa4..5e37d1e62 100644
--- a/include/asio/detail/reactive_socket_service_base.hpp
+++ b/include/asio/detail/reactive_socket_service_base.hpp
@@ -55,6 +55,14 @@ public:
   // The implementation type of the socket.
   struct base_implementation_type
   {
+    // default constructor
+    base_implementation_type()
+      : socket_(invalid_socket),
+        state_(0),
+        reactor_data_{}
+    {
+    }
+
     // The native socket representation.
     socket_type socket_;
 
-- 
2.43.0

